### PR TITLE
Issue 278 log prob parser

### LINKED ISSUE #278: Introduce LogProbParser
Implement the `LogProbParser` component.

## Implementation decision & Insights
This class wraps `parse_top_logprobs` so it can serve as the very first step in the Scikit-Learn `BaseDetector` pipeline, letting users pass raw generator output (OpenAI, vLLM, etc.) directly into `Detector.predict_proba(raw_outputs)`. It is also the **trust boundary** with the (untrusted) API output: it validates the parsed values before emitting a padded numeric array, so every downstream step can assume clean, array-shaped data.

## Output contract
`transform(X) -> np.ndarray` of shape `(n_samples, max_tokens, k)`, dtype float, where `max_tokens` is the longest token sequence in the batch and `k` the number of top logprobs. Returning a dense array (instead of `list[dict[int, list[float]]]`) makes the entropy steps (#239 / #277) trivially vectorizable.

Padding uses **`np.nan`** as the sentinel for absent positions — both ranks missing within a token and tokens missing within a shorter sequence. NaN is safe as a sentinel **only because step 2 below guarantees real data is NaN-free**, so downstream `NaN ⟺ padding` holds by construction (a corrupt/failed API value can never masquerade as padding).

## Algorithm & Steps
1. Create a class `LogProbParser` inheriting from `BaseEstimator` and `TransformerMixin`.
2. **Parse + validate (trust boundary).** Delegate to `parse_top_logprobs` to get the jagged values, then assert every real logprob is finite and in-domain (`logprob <= 0`; not `NaN` / `±inf` / `None`). On any violation, raise a clear `ValueError` identifying the offending position (e.g. `sample i, token j, rank r`). This catches API / parse errors loudly instead of letting them slip through as padding later. (Consistent with the direction of #221 and #225.)
3. **Pad to a dense array.** Stack the validated jagged values into `(n_samples, max_tokens, k)`, filling absent ranks and absent tokens with `np.nan`.
4. Implement a single `transform(X)` method. No `transform_tokens` is needed: the parser performs no token reduction, so `BaseDetector.predict_token_proba` falls back to `transform` for this step; the token-vs-sequence distinction lives in `EntropyTransformer` (#277).
5. Override `__sklearn_tags__(self)` to (a) disable strict 2D input checks (`tags.input_tags.two_d_array = False`, `list_of_dicts = True`) — the parser's *input* is still raw model objects, even though its *output* is an array — and (b) set `tags.requires_fit = False`, since the parser is **stateless** (it learns nothing from data). `fit` just returns `self`, and `transform` works without `fit`.

## Acceptance criteria
- [x] Implements `LogProbParser` as a valid scikit-learn transformer.
- [x] `transform` parses via `parse_top_logprobs` and returns a NaN-padded `(n_samples, max_tokens, k)` float array.
- [x] Validates that all real logprobs are finite and in-domain, raising `ValueError` on non-finite / out-of-domain values *before* padding.
- [x] Overrides `__sklearn_tags__` to accept raw (non-array) input.
- [x] Stateless: `__sklearn_tags__` sets `requires_fit = False`; `transform` works without calling `fit`.

### PR DESCRIPTION
**Pull request by CharlotteLeBihan18**
closes #278

### CODE REVIEW

**Review by chicham — DISMISSED** [review]

File: src/artefactual/preprocessing/parser.py [hunk 1]
```diff
@@ -20,6 +23,55 @@
 )
 
 
+class LogProbParser(BaseEstimator, TransformerMixin):
```
Comment (chicham) [inline src/artefactual/preprocessing/parser.py:26]: **Tests missing.** This adds a new public transformer with no tests. **The snippet below is ready to use as-is** — drop it into [`tests/preprocessing/test_parser.py`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/tests/preprocessing/test_parser.py) to cover `LogProbParser`.

Several behaviors are invariants, so they're written as [`hypothesis`](https://hypothesis.readthedocs.io/) property tests. The generator produces what `transform` actually consumes — a **raw** response shaped the way an OpenAI-compatible API returns logprobs: the **same `top_logprobs` count `k` for every token**, already **sorted descending**, every value finite and `<= 0`. Values come from [`hypothesis.extra.numpy`](https://hypothesis.readthedocs.io/en/latest/numpy.html) as `float32` to match the array `transform` builds:

```python
import numpy as np, pytest
from hypothesis import assume, given, strategies as st
from hypothesis.extra.numpy import arrays

@st.composite
def chat_completion(draw):
    k = draw(st.integers(min_value=1, max_value=20))          # fixed top_logprobs count per response

    def token_entry():
        lps = draw(arrays(dtype=np.float32, shape=k,
                          elements=st.floats(min_value=-30.0, max_value=0.0,
                                             allow_nan=False, allow_infinity=False, width=32)))
        lps = np.sort(lps)[::-1]                               # API returns top_logprobs descending
        return {"top_logprobs": [{"token": "x", "logprob": float(v)} for v in lps]}

    def choice():
        n_tokens = draw(st.integers(min_value=1, max_value=8))
        return {"logprobs": {"content": [token_entry() for _ in range(n_tokens)]}}

    n_choices = draw(st.integers(min_value=1, max_value=3))
    return {"choices": [choice() for _ in range(n_choices)]}


@given(chat_completion())
def test_transform_shape_and_padding(payload):
    arr = LogProbParser().transform(payload)
    assert arr.shape[0] == len(payload["choices"]) and arr.dtype == np.float32
    real = arr[~np.isnan(arr)]
    assert np.isfinite(real).all() and (real <= 0).all()      # NaN <-> padding invariant

@given(chat_completion(), st.data())
def test_transform_rejects_out_of_domain(payload, data):
    c = data.draw(st.integers(0, len(payload["choices"]) - 1))
    content = payload["choices"][c]["logprobs"]["content"]
    assume(content)
    t = data.draw(st.integers(0, len(content) - 1))
    bad = data.draw(st.sampled_from([float("inf"), float("nan"), 1.0]))
    content[t]["top_logprobs"][0]["logprob"] = bad
    with pytest.raises(ValueError, match="Invalid logprob"):
        LogProbParser().transform(payload)
```

Run with `uv run --extra test pytest tests/preprocessing/test_parser.py`. Then add plain `pytest` examples for the rest: **statelessness** (`transform` without `fit`), **sklearn round-trip** (`clone()` / `get_params()` / `set_params()`), and edge cases `[]` → `(0, 0, 0)`, `[{}]` → `(1, 0, 0)`.

Two path-specific notes (OpenAI ChatCompletion route): (1) [`_extract_logprobs_from_token`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/openai_parser.py#L26) drops `None` ([L44](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/openai_parser.py#L44)), so `transform`'s `lp is None` branch is unreachable via a conformant API — a malformed-provider test or a pure `validate_and_pad(parsed)` helper covers it; (2) [`_process_single_choice`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/openai_parser.py#L49) skips empty tokens ([L71](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/openai_parser.py#L71)), so token indices can be non-contiguous — see [`parser.py:53`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/parser.py#L53).

File: src/artefactual/preprocessing/parser.py [hunk 2]
```diff
@@ -20,6 +23,55 @@
...
+    """
+    Wraps parse_top_logprobs as a pipeline step.
+
+    Because this transformer accepts non-standard list[dict] inputs
+    by bypassing scikit-learn validation, cross-validation
+    must start from data that has already been parsed.
+    """
+
+    def fit(self, _x, _y=None) -> "LogProbParser":
```
Comment (chicham) [inline src/artefactual/preprocessing/parser.py]: **Renaming the fit params `_x`/`_y` breaks the sklearn API contract.** sklearn names these `X` and `y`, and callers pass `y` **by keyword**, which `_y` rejects:

```pycon
>>> LogProbParser().fit(payload, y=labels)
TypeError: LogProbParser.fit() got an unexpected keyword argument 'y'
```

Also [`has_fit_parameter`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.has_fit_parameter.html)`(est, "y")` returns `False`, so `y`/`sample_weight` metadata routing can't see it.

Scope: a plain `Pipeline.fit(X, y)` / `cross_validate` pass `y` **positionally**, so they don't raise — the bug is latent there. It surfaces on a direct keyword call `fit(X, y=...)` or a [routed fit-param](https://scikit-learn.org/stable/modules/generated/sklearn.utils.metadata_routing.process_routing.html) (`pipe.fit(X, y, parse__y=...)`). `check_estimator` doesn't guard it either — it's *skipped* because [`__sklearn_tags__`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/parser.py#L63) marks the input non-standard (`SkipTestWarning`). Latent or not, it's a contract violation.

**You most likely used `_x`/`_y` to silence the unused-argument lint** — that's ruff's `ARG002` (enabled via `select = [..., "ARG"]` in `pyproject.toml`. Don't trade the contract for the lint — keep the conventional names and ignore the lint explicitly instead:

```python
def fit(self, X, y=None):  # noqa: ARG002  — X / y unused but required by the sklearn fit signature
    return self
```

([`transform`](https://github.com/artefactory/artefactual/blob/194e9dfa6d4c48ac6db039ff4d25c6db8c511fd6/src/artefactual/preprocessing/parser.py#L38)'s argument is actually used, so renaming `x` → `X` there needs no ignore.) Type hints optional; the rename + ignore is the substantive fix.

File: src/artefactual/preprocessing/parser.py [hunk 3]
```diff
@@ -20,6 +23,55 @@
...
+        # validation step
+        for i, sample in enumerate(parsed):  # sample is the dictionary for each generation
+            for token_idx, logprobs in sample.items():  # token position, ragged list of logprobs
+                for rank, lp in enumerate(logprobs):  # column index or k index, logprob value
+                    if lp is None or not np.isfinite(lp) or lp > 0:
+                        error_msg = (
+                            f"Invalid logprob at sample {i}, token {token_idx}, "
+                            f"rank {rank}: {lp!r}. Expected a finite value <= 0."
+                        )
+                        raise ValueError(error_msg)
```
Comment (chicham) [inline src/artefactual/preprocessing/parser.py:53]: Perf note: this validation is a triple nested Python loop (O(n·T·k)) over every logprob. No need to change it now — but please add a short code comment flagging it as a known hotspot that can be vectorized later (validate the padded array in one shot), e.g.:

```python
# TODO(perf): O(n·T·k) Python loop — vectorize over the padded array once batches get large.
```

**Review by chicham — CHANGES_REQUESTED** [review]
Two items to address before merge: **missing tests** (`:26` — a ready-to-use hypothesis snippet is inline) and the **`_x`/`_y` sklearn-contract break** (`:35` — rename to `X`/`y` + `# noqa: ARG002`). The other two inline notes are a perf marker and an informational `__sklearn_tags__` cross-reference. `LogProbParser` otherwise matches the #278 contract and `ty` passes — clean implementation overall.

File: src/artefactual/preprocessing/parser.py (hunk 2 above)
Comment (CharlotteLeBihan18) [reply src/artefactual/preprocessing/parser.py]: Added # noqa:  N803 comment to fit and transform for ruff to accept that the argument name `X` is not a lowercase.
